### DISCUSSION TITLE
Bun?

### DISCUSSION THREAD

**Discussion by taviks** [tauri-apps/tauri#5837] [General]
Might tauri ever support Bun ([homepage](https://bun.sh/), [github](https://github.com/oven-sh/bun)) as a runtime? 

Any interest in this? 

Anyone know of other projects similar to tauri that use a JS-based runtime (maybe something that compiles to WASM)?

Context (fwiw): I'm a total newbie to tauri and Rust and needing to learn it in order to use tauri feels like learning to walk all over again. However, I've just learned about tauri's JS APIs which might prevent me from needing to do much (if any) Rust (so maybe this is less of an issue than I had thought).

**Comment by ghost** [comment]
Tauri uses browser environment to run js, is there any need for bun?

**Reply by taviks** [reply]
Bun would be used instead of Rust for server/system-level things, not browser-level things. 

Found an Electron feature request for the option to use Bun instead of Node (https://github.com/electron/electron/issues/34876). 

I don't know if there'd be any performance gains or losses, but being able to write server code in JS (instead of Rust) would be a win IMO.

**Reply by ghost** [reply]
Sounds good

**Reply by JonasKruckenberg** [reply]
I can say that we're very aware of the fact Rust has a rather steep learning curve, especially for developers coming from the web.
Allowing people to use a different language to develop their core logic is a _really_ good idea and something we're very interested in enabling.
That said it's not a trivial task and we're currently prioritizing a stable and feature-full API over bindings to other languages and runtimes. But if we do enable JavaScript in the Core expect it to be Deno rather than bun, both because it's written in Rust (and therefore integrates better with Tauri) and because of Denos inherent focus on security.

**Reply by ghost** [reply]
Also all that JavaScript bloat is slow in computing, not memory safe and efficient. Rust is really cool. I remember I had to work with big data, like reading massive files with 3-10M lines each and creating arrays with structs, node could not solve these problems, it was failing always, either stack or something else. The only thing that I don't like about Rust is rust-analyser (vscode extension) 😃  which eats all my RAM

**Comment by FabianLars** [comment]
I guess this would be part of the `Other Bindings` roadmap entry. It's way too early to tell if Bun will ever be part of that, so no promises. It would basically be similar to a sidecar but with a much tighter integration. And Rust will always be a part of Tauri so it wouldn't be a complete switch but just bolted on top.

> However, I've just learned about tauri's JS APIs which might prevent me from needing to do much (if any) Rust

That's the idea behind the js api (and to an extend the plugin system). Tauri is still quite young so the api doesn't cover all the common use cases yet

**Reply by taviks** [reply]
> That's the idea behind the js api (and to an extend the plugin system).

Learning about this feature of Tauri was a revelation of sorts. "Oh, I don't have to write "server" code for reading from and writing to the filesystem(!)". It's quite a nice design. Thank you all! Bravo!

> Tauri is still quite young so the api doesn't cover all the common use cases yet

Yes, this is good to keep in mind. 

One such usecase I've hit (not sure how common it is) is needing to find the paths of all occurances of a specific file (e.g. all `README.md` files, at least under the user's "home" directory). It wasn't too hard to create my own Rust function that does this, though I'm unsure of how sound it is (feedback welcome):

```rust
#[tauri::command]
async fn get_all_readme_files() -> Result<Vec<String>, String> {
  use rust_search::SearchBuilder;
  
  let search: Vec<String> = SearchBuilder::default()
    .location("~")
    .search_input("README")
    .limit(100) // results to return
    .ext("md")
    .strict()
    .depth(10)
    .hidden()
    .build()
    .collect();

  return Ok(search);
}
```

I could see good value in enhancing the existing https://tauri.app/v1/api/js/fs#readdir function to accept a glob (or have some other way of narrowing the results, and excluding matches to help with performance), but there would be other ways to approach this and maybe something along these lines is already planned.

**Reply by techpeace** [reply]
Another approach (referenced above) is [using Tauri’s sidecar functionality with `pkg` to bundle a node server with your app](https://tauri.app/v1/guides/building/sidecar/#using-nodejs-on-a-sidecar). There’s an example listed in the linked doc, and [this project](https://github.com/Perfect7M/taurine) takes the same approach.

**Comment by xueweiwujxw** [comment]
If you want to use bun to create and develop, you can just use `bun create tauri-app`. 
And replace `beforeDevCommand`'s value and `beforeBuildCommand`'s value in `tauri.config.json` with `bun dev` and `bun build`. 
```json
"build": {
  "beforeDevCommand": "bun dev",
  "beforeBuildCommand": "bun build",
  "devPath": "http://localhost:1420",
  "distDir": "../dist",
  "withGlobalTauri": false
},
```
I tried it and it works, but can't guarantee if there are other potential issues.

**Reply by xueweiwujxw** [reply]
Maybe Tauri can add `bun` into package manager choices.

**Reply by xueweiwujxw** [reply]
Oh Wait. `bun build` has  some issues.

**Reply by xueweiwujxw** [reply]
This configs can work.
tauri.config.json
```json
  "build": {
    "beforeDevCommand": "bun run dev",
    "beforeBuildCommand": "bun run build",
    "devPath": "http://localhost:1420",
    "distDir": "../dist",
    "withGlobalTauri": false
  },
```
package.json
```json
  "scripts": {
    "dev": "bunx --bun vite",
    "build": "bunx --bun vite build",
    "preview": "vite preview",
    "tauri": "tauri"
  },
```
You can run `bun run dev`, `bun run build`, `bun tauri dev` and `bun tauri build`

**Reply by s9kt** [reply]
I was talking to someone about this and `bun --bun run tauri` doesn't appear to work, I get a strict ESM error:

`error: Declarations with the name arguments cannot be used with esm due to strict mode`

May be a bun problem but just be aware that without the `--bun` flag, this is [likely running Node.js rather than bun](https://bun.sh/docs/cli/run#bun)

**Reply by xueweiwujxw** [reply]
I tried this - [Build a frontend using Vite and Bun](https://bun.sh/guides/ecosystem/vite), and run `bun tauri build`

**Comment by Nishchit14** [comment]
This is still relevant, is there any planning to support a js runtime? bun/deno/node

rust is great but having a js runtime would help in sharing the common js resources with web and backend like types, services, libs, etc.

**Comment by Tyjust31** [comment]
we can't with the community with open source elements of sracht try to have a desktop application with bun and sveltkit or whatever i know it's complicated but it concerns me

**Comment by trymeouteh** [comment]
I would like to see a node like backend option for tauri that is light. Would rather use JS in the backend over rust and I think the developer should be able to choose weather to use rust or a light js backend.

**Reply by 3commascapital** [reply]
it would be cool to at least be able to choose an executable to be included and more easily run sidecars

**Comment by joodaloop** [comment]
+1 for adding an optional Node.js/other JS runtime backend, it might only increase bundle size/memory usage a little bit compared to Electron's full Chromium bundle, and would enable much faster development for many people.

**Reply by joodaloop** [reply]
Guess I was right, Electrobun is growing fast a few weeks after launch and will be what I use if I need a desktop app with a webview.

https://github.com/blackboardsh/electrobun

**Comment by a-eid** [comment]
this is probably the best decision tauri team can make.

**Comment by Gin-Quin** [comment]
Yep, Tauri would get all JS developers that are afraid of Rust—which is a lot. And Bun is a very sexy tech right now.

**Reply by Nishchit14** [reply]
who is listening?

**Comment by Hendrixer** [comment]
Cloudflare v8 isolsates would be perfect for this imo

**Comment by zbycz** [comment]
Someone published an article+repo with Bun as tauri sidecar. I haven't personaly tested it yet.

https://codeforreal.com/blogs/using-bun-or-deno-as-a-web-server-in-tauri/
